[develop] First implementation of run_WE2E_tests.py

mkavulich opened the pull request:
## DESCRIPTION OF CHANGES: 
This PR introduces two new scripts to the repository: `run_WE2E_tests.py` and `monitor_jobs.py`. The purpose of these scripts is to eventually provide a pythonic replacement for the current workflow end-to-end test submission script. Additionally, the monitor_jobs function gives the capability to monitor and submit jobs automatically via the command line or a batch job, rather than relying on crontab entries.

### run_WE2E_tests.py

This new script is roughly analogous to the legacy bash version, `run_WE2E_tests.sh`. This script will set up and run a set of workflow end-to-end tests as specified by the user. 

This script is similar in behavior to the original script, but introduces several improvements and/or simplifications:

 - The argument list is made simpler and more pythonic (which will solve problems listed in #369)
 - The three various test specification arguments (`tests_file`, `test_type`, and `test_name`) are collapsed into a single argument, `--tests`. 
 - The large bit of logic assessing available tests, linking test names, IDs, and subdirectories is omitted; replaced with simple checks based on file names. 
 - "Verbose" and "Debug" options as applied to individual tests are replaced with "verbose_tests" and "debug_tests" respectively, and a "--debug" flag for script verbosity is added.
 - A new "--quiet" flag is introduced, which prevents output (except for warnings and errors) from generate_FV3LAM_workflow printing to screen (they will still be printed to the log file, `log.run_WE2E_tests`). This can prevent important messages from being drowned out by the flood of information printed to screen.
 - Because we are not intermingling all variables from test config, defaults, and machine defaults together, we can better control the resulting test files that are submitted to the generate function
 - Omits checks that are redundant to checks in generate_FV3LAM_workflow.py (for example, ensuring the forecast length is evenly divisible by the LBC interval)
 - Unless `use_cron_to_relaunch` is set to true, the experiment data will be automatically fed into the `monitor_jobs()` function, which will launch and track experiments until all are complete. If the script is interrupted for some reason, it can be re-run and experiments continued (see below)

#### Example usage:
 - Running a list of tests in a file named "testlist" on Hera
`./run_WE2E_tests.py -m=hera -a=fv3lam --tests=testlist`
 - Running the single test "nco_inline_post" on orion, while not printing output from generate_FV3LAM_workflow to screen
`./run_WE2E_tests.py -t=nco_inline_post -m=orion -a=gsd-fv3-test -q`
 - Running the "comprehensive" set of tests on Jet, in debug mode (lots more output print to screen)
`./run_WE2E_tests.py -m=jet -a=gsd-fv3-dev --tests=comprehensive -d`
 - Running the "fundamental" set of tests on Cheyenne, with the gnu compiler, with all experiment directories under the directory "/glade/scratch/kavulich"
`./run_WE2E_tests.py -t=fundamental -c=gnu -m=cheyenne -a=P48500053 --expt_basedir=/glade/scratch/kavulich`

#### Example output:
The `--help` flag gives some usage information:

```
$ ./run_WE2E_tests.py -h
usage: run_WE2E_tests.py [-h] -m MACHINE -a ACCOUNT -t [TESTS ...] [-c COMPILER] [-d] [-q] [--modulefile MODULEFILE]
                         [--run_envir RUN_ENVIR] [--expt_basedir EXPT_BASEDIR] [--exec_subdir EXEC_SUBDIR]
                         [--use_cron_to_relaunch] [--cron_relaunch_intvl_mnts CRON_RELAUNCH_INTVL_MNTS] [--debug_tests]
                         [--verbose_tests]

required arguments:
  -m MACHINE, --machine MACHINE
                        Machine name; see ush/machine/ for valid values
  -a ACCOUNT, --account ACCOUNT
                        Account name for running submitted jobs
  -t [TESTS ...], --tests [TESTS ...]
                        Can be one of three options (in order of priority): 1. A test name or list of test names. 2. A
                        test suite name ("fundamental", "comprehensive", or "all") 3. The name of a file (full or relative
                        path) containing a list of test names.

optional arguments:
  -h, --help            show this help message and exit
  -c COMPILER, --compiler COMPILER
                        Compiler used for building the app
  -d, --debug           Script will be run in debug mode with more verbose output
  -q, --quiet           Suppress console output from workflow generation; this will helpkeep the screen uncluttered
  --modulefile MODULEFILE
                        Modulefile used for building the app
  --run_envir RUN_ENVIR
                        Overrides RUN_ENVIR variable to a new value ( "nco" or "community" ) for all experiments
  --expt_basedir EXPT_BASEDIR
                        Explicitly set EXPT_BASEDIR for all experiments
  --exec_subdir EXEC_SUBDIR
                        Explicitly set EXEC_SUBDIR for all experiments
  --use_cron_to_relaunch
                        Explicitly set USE_CRON_TO_RELAUNCH for all experiments; this option disables the "monitor" script
                        functionality
  --cron_relaunch_intvl_mnts CRON_RELAUNCH_INTVL_MNTS
                        Overrides CRON_RELAUNCH_INTVL_MNTS for all experiments
  --debug_tests         Explicitly set DEBUG=TRUE for all experiments
  --verbose_tests       Explicitly set VERBOSE=TRUE for all experiments

For more information about config arguments (denoted in CAPS), see ush/config_defaults.yaml
```

For a real example of its usage, here I use the `-q` flag to suppress output from `generate_FV3LAM_workflow()`, showing just the output being made by this script:

```
./run_WE2E_tests.py -t=fundamental -m=orion -a=gsd-fv3-test -q 
Checking that all tests are valid
Will run 5 tests:
/work/noaa/gsd-fv3-test/kavulich/UFS/issue_462/ufs-srweather-app/tests/WE2E/test_configs/grids_extrn_mdls_suites_community/config.grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_RRFS_v1beta.yaml
/work/noaa/gsd-fv3-test/kavulich/UFS/issue_462/ufs-srweather-app/tests/WE2E/test_configs/grids_extrn_mdls_suites_community/config.grid_SUBCONUS_Ind_3km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta.yaml
/work/noaa/gsd-fv3-test/kavulich/UFS/issue_462/ufs-srweather-app/tests/WE2E/test_configs/wflow_features/config.community_ensemble_008mems.yaml
/work/noaa/gsd-fv3-test/kavulich/UFS/issue_462/ufs-srweather-app/tests/WE2E/test_configs/wflow_features/config.deactivate_tasks.yaml
/work/noaa/gsd-fv3-test/kavulich/UFS/issue_462/ufs-srweather-app/tests/WE2E/test_configs/wflow_features/config.inline_post.yaml

Inline post is turned on, deactivating post-processing tasks:
RUN_TASK_RUN_POST = False

calling function that monitors jobs, prints summary
Writing information for all experiments to monitor_jobs_20230124154853.yaml
Checking tests available for monitoring...
Starting experiment grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_RRFS_v1beta running
Starting experiment grid_SUBCONUS_Ind_3km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta running
Starting experiment community_ensemble_008mems running
Starting experiment deactivate_tasks running
Starting experiment inline_post running
Setup complete; monitoring 5 experiments
Experiment deactivate_tasks is COMPLETE; will no longer monitor.
Experiment grid_SUBCONUS_Ind_3km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta is COMPLETE; will no longer monitor.
Experiment grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_RRFS_v1beta is COMPLETE; will no longer monitor.
Experiment inline_post is COMPLETE; will no longer monitor.
Experiment community_ensemble_008mems is COMPLETE; will no longer monitor.
All 5 experiments finished in 0:46:43.495644
All experiments are complete
Summary of results available in monitor_jobs_20230124154853.yaml
```

Note that the message about Inline post is a "warning"-level message from generate_FV3LAM_workflow, and so is still printed to screen.
### monitor_jobs.py

This new script, designed to be called automatically by run_WE2E_tests.py or run stand-alone, will read a dictionary (either provided directly to the function or read from a YAML file) that specifies the location of a number of experiments that need to be monitored. The main function, monitor_jobs(), will keep track of these experiments, advance the workflow with calls to `rocotorun` in each experiment directory, and monitor successes and errors as they occur, reporting a summary at the end.

In addition, while these jobs are being monitored and run, a YAML file tracking the status of all jobs will be written to disk. This file can be read directly to see the details of how each job is coming along, but most importantly, this job file can be read back into monitor_jobs.py as a command line argument. Therefore if the `./run_WE2E_tests.py` script fails or is quit at any point after the experiments have been generated, the script can be re-started and continue to monitor jobs where it left off.

#### Example usage:
In this case, I ran `./run_WE2E_tests.py` but killed it before all experiments completed. Here I just look in my test directory for the latest "monitor_jobs" yaml file, and feed that back to the script:
```
$ ls -l
total 488
-rw-r--r-- 1 Michael.Kavulich fv3lam   5827 Jan 17 15:59 create_WE2E_resource_summary.py
-rwxr-xr-x 1 Michael.Kavulich fv3lam  15347 Jan 17 15:59 get_expts_status.sh
-rwxr-xr-x 1 Michael.Kavulich fv3lam  64794 Jan 17 15:59 get_WE2Etest_names_subdirs_descs.sh
-rw-r--r-- 1 Michael.Kavulich fv3lam  12423 Jan 23 21:33 log.monitor_jobs
-rw-r--r-- 1 Michael.Kavulich fv3lam 182013 Jan 24 21:55 log.run_WE2E_tests
drwxr-sr-x 2 Michael.Kavulich fv3lam   4096 Jan 23 22:19 machine_suites
-rw-r--r-- 1 Michael.Kavulich fv3lam    915 Jan 24 21:55 monitor_jobs_20230124215513.yaml
-rwxr-xr-x 1 Michael.Kavulich fv3lam  12508 Jan 24 00:26 monitor_jobs.py
drwxr-sr-x 2 Michael.Kavulich fv3lam   4096 Jan 24 18:47 __pycache__
drwxr-sr-x 9 Michael.Kavulich fv3lam   4096 Jan 20 04:51 rocoto
-rwxr-xr-x 1 Michael.Kavulich fv3lam  22882 Jan 24 21:50 run_WE2E_tests.py
-rwxr-xr-x 1 Michael.Kavulich fv3lam  50269 Jan 17 16:00 run_WE2E_tests.sh
-rwxr-xr-x 1 Michael.Kavulich fv3lam   2844 Jan 17 15:59 setup_WE2E_tests.sh
drwxr-sr-x 6 Michael.Kavulich fv3lam   4096 Jan 23 16:58 test_configs
$ cat monitor_jobs_20230124222802.yaml 
### WARNING ###
### THIS FILE IS AUTO_GENERATED AND REGULARLY OVER-WRITTEN BY monitor_jobs.py
### EDITS MAY RESULT IN MISBEHAVIOR OF EXPERIMENTS RUNNING
deactivate_tasks:
  expt_dir: /scratch2/BMC/fv3lam/kavulich/UFS/workdir/issue_462_new_WE2E_script/expt_dirs/deactivate_tasks
  status: CREATED
nco_grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR:
  expt_dir: /scratch2/BMC/fv3lam/kavulich/UFS/workdir/issue_462_new_WE2E_script/expt_dirs/nco_grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
  status: CREATED
custom_ESGgrid:
  expt_dir: /scratch2/BMC/fv3lam/kavulich/UFS/workdir/issue_462_new_WE2E_script/expt_dirs/custom_ESGgrid
  status: CREATED
$ ./monitor_jobs.py -y=monitor_jobs_20230124215513.yaml 
Checking tests available for monitoring...
Starting experiment deactivate_tasks running
Starting experiment nco_grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR running
Starting experiment custom_ESGgrid running
Setup complete; monitoring 3 experiments
Experiment deactivate_tasks is COMPLETE; will no longer monitor.
Experiment nco_grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR is COMPLETE; will no longer monitor.
Experiment custom_ESGgrid is COMPLETE; will no longer monitor.
All 3 experiments finished in 0:15:59.928853
```

And querying the monitor file afterwards shows that all experiments are indeed complete:
```
$ cat monitor_jobs_20230124215513.yaml 
### WARNING ###
### THIS FILE IS AUTO_GENERATED AND REGULARLY OVER-WRITTEN BY monitor_jobs.py
### EDITS MAY RESULT IN MISBEHAVIOR OF EXPERIMENTS RUNNING
deactivate_tasks:
  expt_dir: /scratch2/BMC/fv3lam/kavulich/UFS/workdir/issue_462_new_WE2E_script/expt_dirs/deactivate_tasks
  status: COMPLETE
  make_grid_201907010000: SUCCEEDED
  make_orog_201907010000: SUCCEEDED
  make_sfc_climo_201907010000: SUCCEEDED
nco_grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR:
  expt_dir: /scratch2/BMC/fv3lam/kavulich/UFS/workdir/issue_462_new_WE2E_script/expt_dirs/nco_grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
  status: COMPLETE
  get_extrn_ics_202008100000: SUCCEEDED
  get_extrn_lbcs_202008100000: SUCCEEDED
  make_ics_202008100000: SUCCEEDED
  make_lbcs_202008100000: SUCCEEDED
  run_fcst_202008100000: SUCCEEDED
custom_ESGgrid:
  expt_dir: /scratch2/BMC/fv3lam/kavulich/UFS/workdir/issue_462_new_WE2E_script/expt_dirs/custom_ESGgrid
  status: COMPLETE
  make_grid_201907010000: SUCCEEDED
  get_extrn_ics_201907010000: SUCCEEDED
  get_extrn_lbcs_201907010000: SUCCEEDED
  make_orog_201907010000: SUCCEEDED
  make_sfc_climo_201907010000: SUCCEEDED
  make_ics_201907010000: SUCCEEDED
  make_lbcs_201907010000: SUCCEEDED
  run_fcst_201907010000: SUCCEEDED
  run_post_f000_201907010000: SUCCEEDED
  run_post_f001_201907010000: SUCCEEDED
  run_post_f002_201907010000: SUCCEEDED
  run_post_f003_201907010000: SUCCEEDED
  run_post_f004_201907010000: SUCCEEDED
  run_post_f005_201907010000: SUCCEEDED
  run_post_f006_201907010000: SUCCEEDED
```

### Additional changes

In addition to the new scripts, a few changes have been made to the rest of the workflow that should have no significant impact on existing tests:

 - generate_FV3LAM_wflow.py
    - function generate_FV3LAM_wflow() now takes an optional "debug" argument that, when true, will provide more verbose output from this and other functions that receive the argument.
    - function generate_FV3LAM_wflow() now returns EXPTDIR, the string indicating the full path of the created experiment. This is useful for scripts/functions that call this function directly rather than via command-line
    - Move the "Experiment generation completed" message to the *very* end of the script, so that any failure will pre-empt it
    - Generalize logging setup so that it will work properly even if called from another function.

 - setup.py
   - function setup() now takes an optional "debug" argument that, when true, will provide more verbose output
   - Add several debug logging messages, remove an old "print" statement
   - Avoid using "cd" and other full-path changes so that results will be the same regardless of the directory this function is called from

 - tests/WE2E/test_configs/wflow_features/config.deactivate_tasks.yaml
   - Remove sections for tasks that are not used (LBCS and ICS are not staged since those tasks are deactivated

### Missing features vs. run_WE2E_tests.sh

This initial version is incomplete, and a few more capabilities must be added before a wholesale replacement of run_WE2E_tests.sh

 - The command-line options `cron_relaunch_intvl_mnts`, `generate_csv_file`, and `opsroot` are not yet implemented
 - Currently, the default behavior is to *not* submit jobs to cron, but rather to use monitor_jobs.py to monitor and run the experiments
 - The ability to include calls to UNIX `date` in DATE_FIRST_CYCL and DATE_LAST_CYCL is not yet implemented
 - No checks that all MET executables/directories are set up properly for the given machine...should these be implemented in setup.py instead?
 - No hard-coded "maxtries" based on platform yet
 - Potentially other edge use cases I have missed?

### Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## TESTS CONDUCTED: 

- [x] hera.intel
- [x] orion.intel
- [x] cheyenne.intel
- [x] cheyenne.gnu
- [ ] gaea.intel
- [x] jet.intel
- [ ] wcoss2.intel
- [ ] NOAA Cloud (indicate which platform)
- [ ] Jenkins
- [x] fundamental test suite
- [x] comprehensive tests (specify *which* if a subset was used)
    - [x] Jet

## DEPENDENCIES:
None

## DOCUMENTATION:
Will be contributed with later PR deprecating old system. This is a preliminary implementation for wider exposure and feedback.

## ISSUE: 
 - This is partially in response to #462. Still needs some more work before it can be closed.
 - When this script supersedes the legacy bash script, it will also solve #369
 - Some changes I had to make here (related to calling the generate function from another directory) also fix #539

## CHECKLIST
<!-- Add an X to check off a box. -->
- [ ] My code follows the style guidelines in the Contributor's Guide
- [ ] I have performed a self-review of my own code using the Code Reviewer's Guide
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes do not require updates to the documentation (explain).
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

mkavulich committed:
Initial version, takes file with list of tests and checks they are all valid

mkavulich committed:
Add argparse

mkavulich committed:
- Add more command line options, format help message better
 - Write new config file and print as test

mkavulich committed:
Use direct calls to logging, use args.debug and logging.debug to manage what gets printed to screen, and have all messages printed to log file

mkavulich committed:
Add more command line arguments from original script, update config.yaml accordingly

mkavulich committed:
Almost ready to call generate_FV3LAM_wflow! Added some section-specific functions, not ready to use yet

mkavulich committed:
Implemented test specification as a single, unified command-line argument. Much cleaner implementation

mkavulich committed:
SUCCESS! Script now runs on a card-coded test input file. Next step is
implement rest of the command line options, then finally implement
non-cron submission and tracking of jobs. Changes in this commit:

 - Ensure all dictionary entries within each section are uppercase
 - Implement functions for updating settings in task_get_extrn_ics and
   task_get_extrn_lbcs sections
 - Import generate_FV3LAM_wflow function directly and call it
 - Implement debigging arguments in generate_FV3LAM_wflow.py and
   setup.py
 - Tweak setup_logging to avoid double-printing of log messages when
   calling generate_FV3LAM_wflow from another function with logging

mkavulich committed:
Fix test filename bug

mkavulich committed:
Updates to logging to ensure log files are written correctly and in the
correct location.

mkavulich committed:
Improvements to workflow generation

 - Remove workdir dependencies from setup function
 - Return "EXPT_DIR" from workflow generation function, move "completion"
message to outside of generation function to avoid confusing/incorrect
"success" messages

mkavulich committed:
Finish up initial version of run_WE2E_tests.py, start work on
monitor_jobs.py (currently just starts the first rocotorun of each
experiment)

mkavulich committed:
Continue to populate monitor_jobs.py

 - Add function for writing monitor file (this will be overwritten each
time)
 - Add finction to query rocoto database file for each experiment (using
sqlite3), and extract the relevant information for each job
 - Add a skeleton function for updating the status of each experiment
(based on each job's individual status within that experiment)

mkavulich committed:
monitor_jobs.py nearly complete!

mkavulich committed:
If RUN_TASK_GET_EXTRN_ICS and/or RUN_TASK_GET_EXTRN_LBCS are false, do not fail if the respective input files don't exist (since that task will not be run anyway)

mkavulich committed:
Ladies and Gentlemen and Smizmars...we have a working prototype!

monitor_jobs.py:
 - Remove verbose flags from rocotorun calls, since we don't use the
output anyway
 - Remove some debug prints, add some others
 - We need to loop over a copy of running_expts rather than the original
so we can remove entries within the loop
 - Move rocotorun to the end of the status check loop to give time for
the rocoto database to be fully updated before it is checked again
 - Add a short delay between loops over running_expts for further safety
against slowly updating databases

run_WE2E_tests.py:
 - If RUN_TASK_GET_EXTRN_ICS or RUN_TASK_GET_EXTRN_LBCS are False, skip
the setup function for that respective task
 - Fix staged paths for ICS and LBCS

mkavulich committed:
Fixing bug in specify_template_filenames test: files were re-named in setup.py but not in test config file

mkavulich committed:
Address lists correctly so --tests argument works correctly

mkavulich committed:
Revert "If RUN_TASK_GET_EXTRN_ICS and/or RUN_TASK_GET_EXTRN_LBCS are false, do not fail if the respective input files don't exist (since that task will not be run anyway)"

I forgot that the default for USE_USER_STAGED_EXTRN_FILES is false, so if True then we should expect failure.

This reverts commit ef072f990996daa7fd5925b6256d24d11ece1105.

mkavulich committed:
Remove sections "task_get_extrn_ics" and "task_get_extrn_lbcs" from config.deactivate_tasks.yaml; since we are not running these tasks, we should not set any of those settings

mkavulich committed:
In monitor_jobs, need to make a copy of running_expts explicitly ( = operator simply references the original). Also add some more debugging/timing info.

mkavulich committed:
Script now runs most of Cheyenne's fundamental tests...still need to figure out what to do with specify_template_filenames...

mkavulich committed:
Make things a bit more verbose, especially for debug mode

mkavulich committed:
Implement command-line submission of monitor_jobs.py. This is
particularly handy for debugging the functionality, and seems to work
flawlessly as implemented :)

Additionally, adding multiple calls to rocotorun in a row to get around
a potential bug with rocotorun leaving hung background processes. In
correspondance with Chris to try to solve this in a cleaner way.

mkavulich committed:
Had more discussion with Chris Harrop about nature of rocotorun issue; seems to be specific to Hera head nodes (but could appear elsewhere)

mkavulich committed:
Improvements to run_WE2E_tests.py

 - Remove tests that are symlinks to tests already included
 - Fix bug in capability to include list of tests as a file
 - Remove some unnecessary prints

mkavulich committed:
More sprucing up of the new script

 - Handle blank/empty lines in a test file without failure
 - Omit duplicate tests similarly to symlink duplicates

mkavulich committed:
add 'quiet' option to suppress test generation output

mkavulich committed:
Remove some old debug prints

mkavulich committed:
Cleanup from pylint

mkavulich committed:
Forgot to add example monitor_jobs.yaml

MichaelLueken reviewed (approved):
@mkavulich Thank you very much for creating the PR to bring in the initial implementation of the pythonized run_WE2E_tests.sh script!  I was able to run a manual test to ensure that Jenkins tests would still work, then submitted the fundamental tests using the new run_WE2E_tests.py script.  Everything worked like a charm!  Approving these changes.

danielabdi-noaa commented on ush/generate_FV3LAM_wflow.py:
```diff
@@ -744,21 +731,36 @@ def generate_FV3LAM_wflow(ushdir, logfile: str = "log.generate_FV3LAM_wflow") ->
+    logging.getLogger().addHandler(fh)
     logging.debug(f"Finished setting up debug file logging in {logfile}")
+
+    # If there are already multiple handlers, that means generate_FV3LAM_workflow was called from another function.
+    # In that case, do not change the console (print-to-screen) logging.
+    if len(logging.getLogger().handlers) > 1:
+        return
```
I am curious, was this what solved the unittest failures?

mkavulich replied:
These changes in `setup_logging()` were needed to get log files (and messages) to behave correctly when called from another function with logging, so if that's how the unit tests are being called then I bet these changes as a whole are what fixed them. Specifically, the use of `logging.basicConfig` was the issue, as it has no effect when there already exists a logging handler. This could be solved with some settings that are only available in Python 3.8+, but this new implementation is cleaner anyway IMO.

MichaelLueken commented:
@mkavulich The Jenkins test, `specify_template_filenames`, on Cheyenne Intel, is failing with the following in the run_fcst step:

```
NOTE from PE     0: diag_manager_mod::diag_manager_init: diag_manager is using fms2_io

FATAL from PE     0: get_ascii_file_num_lines: File diag_table does not exist.


FATAL from PE     0: get_ascii_file_num_lines: File diag_table does not exist.

Image              PC                Routine            Line        Source
ufs_model          000000000432EFB7  Unknown               Unknown  Unknown
ufs_model          000000000274B6E9  mpp_mod_mp_mpp_er          70  mpp_util_mpi.inc
ufs_model          0000000002742A81  mpp_mod_mp_get_as        1476  mpp_util.inc
ufs_model          0000000002612E75  fms_io_utils_mod_         507  fms_io_utils.F90
```
Please see `/glade/scratch/epicufsrt/jenkins/workspace/fs-srweather-app_pipeline_PR-558__2/expt_dirs/specify_template_filenames` for more details.

gsketefian commented:
@mkavulich One thing I noticed with the current (bash) version of `run_WE2E_tests.sh` is that if there is an entry in the test config yaml file (`config.<test_name>.yaml`) that is not defined in `config_defaults.yaml`, it is not flagged by the script but simply ignored.  Does the new python version `run_WE2E_tests.py` in this PR flag such entries (haven't gotten to trying it myself yet)?  Thanks.

mkavulich committed:
Merge remote-tracking branch 'origin/develop' into feature/issue_462_new_WE2E_script_rebased

MichaelLueken commented on tests/WE2E/test_configs/wflow_features/config.specify_template_filenames.yaml:
```diff
@@ -5,11 +5,11 @@ metadata:
-  NEMS_CONFIG_TMPL_FN: nems.configure
+  DATA_TABLE_FN: data_table
+  DIAG_TABLE_FN: diag_table.FV3_GFS_v15p2
+  FIELD_TABLE_FN: field_table.FV3_GFS_v15p2
```
Since these two table indices include .{{ CCPP_PHYS_SUITE }}, shouldn't they be DIAG_TABLE_TMPL_FN and FIELD_TABLE_TMPL_FN?  Making this change on Cheyenne and manually rerunning the specify_template_filenames test allowed the test to run successfully.

mkavulich replied:
@MichaelLueken This was related to the problem I mentioned yesterday: these variables are not named correctly in develop, and that was causing the new script to fail. However, this fix needs to be applied more carefully than this, so I have reverted these changes for now in order to avoid the failure you pointed out. I will open an issue later with more details to hopefully fix this.

MichaelLueken replied:
@mkavulich Thank you very much for the explanation and reverting this change for now.  I certainly agree that the entries for `DATA_TABLE_TMPL_FN`, `MODEL_CONFIG_TMPL_FN`, and `NEMS_CONFIG_TMPL_FN` appear to be set incorrectly with respect to the `specify_template_filenames` test.  As a matter of fact, there are no entries for either `MODEL_CONFIG_TMPL_FN` or `NEMS_CONFIG_TMPL_FN` in `config_defaults.yaml` and `data_table` would need to be changed to `data_table.FV3_GFS_v15p2` in order to properly be set to `DATA_TABLE_TMPL_FN`, so I can agree that these would likely cause the new script to fail.  I'll resolve this conversion for now, reapprove, and then rerun the Cheyenne tests.

MichaelLueken reviewed (changes requested):
@mkavulich After a little digging, I think I discovered the issue that is causing the failure in the specify_template_filenames test for Cheyenne Intel.  From the ush/config_defaults.yaml file, if either the diag_table or field_table include .{{ CCPP_PHYS_SUITE }}, then it should be named either DIAG_TABLE_TMPL_FN or FIELD_TABLE_TMPL_FN, respectively.  Making this change and manually rerunning the specify_template_filenames test showed that it ran successfully.

mkavulich committed:
Revert changes to config.specify_template_filenames.yaml. This will stop the test from failing for the current run_WE2E_tests.sh; but this issue still needs to be addressed as this test currently does not test what it is supposed to

christinaholtNOAA commented on tests/WE2E/monitor_jobs.yaml:
```diff
@@ -0,0 +1,54 @@
+# This is an example yaml file showing the various entries that can be created for tracking jobs by monitor_jobs.py
```
How should this file be updated when changes are made to monitor_jobs.py such that this file is modified?

Can that be a GitHub action?

mkavulich replied:
It is unclear to me how that would be handled in an automatic way. Is it worth the effort for something that will likely not change very often, if at all?

christinaholtNOAA replied:
It could be nice just to drop in a comment on how this appeared here so that future you is not completely baffled. ;)

christinaholtNOAA commented on tests/WE2E/monitor_jobs.py:
```diff
@@ -0,0 +1,280 @@
+        logging.info(f"Starting experiment {expt} running")
+        num_expts += 1
+        rocoto_db = f"{expt_dict[expt]['expt_dir']}/FV3LAM_wflow.db"
+        subprocess.run(["rocotorun", f"-w {expt_dict[expt]['expt_dir']}/FV3LAM_wflow.xml", f"-d {rocoto_db}"])
```
Why not similarly define `rocoto_xml` for consistency?

mkavulich replied:
I did you one better and defined the whole command passed to `subprocess.run`, since that is always the same.

christinaholtNOAA commented on tests/WE2E/monitor_jobs.py:
```diff
@@ -0,0 +1,280 @@
+            # For each entry from rocoto database, store that under a dictionary key named TASKNAME_CYCLE
+            # Cycle comes from the database in Unix Time (seconds), so convert to human-readable
+            cycle = datetime.utcfromtimestamp(task[1]).strftime('%Y%m%d%H%M')
+            expt_dict[expt][f"{task[0]}_{cycle}"] = task[2]
```
`task[2]` here depends on your `SELECT` statement never changing. It might be helpful for the user if you made a named tuple to hold the information in the task so that you don't have to worry about order here and losing track.

mkavulich replied:
I do not see a way to convert the output from fetchall() to a list of namedtuples without significantly increasing the complexity of the code. I'd rather just handle this with a comment.

christinaholtNOAA commented on tests/WE2E/monitor_jobs.py:
```diff
@@ -0,0 +1,280 @@
+            expt_dict[expt][f"{task[0]}_{cycle}"] = task[2]
+        expt_dict[expt] = update_expt_status(expt_dict[expt], expt)
+        #Run rocotorun again to get around rocotobqserver proliferation issue
+        subprocess.run(["rocotorun", f"-w {expt_dict[expt]['expt_dir']}/FV3LAM_wflow.xml", f"-d {rocoto_db}"])
```
Will the calls to subprocess like this wait until `rocotorun` completes before control is returned to the Python script? If not, you might be over-taxing the system here.

Also, is `rocotorun` also running in a cron job at this point?

mkavulich replied:
`subprocess.run` waits for the subprocess to complete before continuing ([python documentation here](https://docs.python.org/3/library/subprocess.html#using-the-subprocess-module)). I decided not to try to capture the output from the subprocess because this seems to slow things down drastically (like, orders of magnitude slower).

The `monitor_jobs()` function is only called by `run_WE2E_tests.py` if the `--use_cron_to_relaunch` option is not specified, so this should never be run in conjunction with crontab running. But I can't control what users choose to do; theoretically this should work just fine if someone also had a crontab entry for this job and called this script manually.

christinaholtNOAA commented on tests/WE2E/monitor_jobs.py:
```diff
@@ -0,0 +1,280 @@
+    logging.info(f'Setup complete; monitoring {num_expts} experiments')
+
+    #Make a copy of experiment dictionary; will use this copy to monitor active experiments
+    running_expts = expt_dict.copy()
```
Do you need a deep copy here? Python docs say this is a shallow copy, so you might be updating `expt_dict` when you write to `running_expts`

mkavulich replied:
From my reading of [the python docs](https://docs.python.org/3/library/copy.html) that should not be necessary. The logic in this loop does not change any of the individual entries to the list (edit: in a way that will not also be reflected in the list copy; it could be that the line `running_expts[expt] = expt_dict[expt]` is redundant), it merely pops each entry out of the copy of the list as the corresponding test completes. A shallow copy should be sufficient for this purpose.

christinaholtNOAA commented on tests/WE2E/monitor_jobs.py:
```diff
@@ -0,0 +1,280 @@
+    while running_expts:
+        i += 1
+        for expt in running_expts.copy():
+            logging.debug(f"Updating status of {expt}")
```
This seems awfully similar to the code above. Perhaps a function would be helpful to reduce duplication?

mkavulich replied:
I thought I commented here but I guess it got lost. I updated the `update_expt_status()` function to consolidate more duplicate code.

christinaholtNOAA commented on tests/WE2E/monitor_jobs.py:
```diff
@@ -0,0 +1,280 @@
+
+    if "DEAD" in statuses:
+        if ( "RUNNING" in statuses ) or ( "SUBMITTING" in statuses ) or ( "QUEUED" in statuses ):
+            expt["status"] = "DYING"
```
This is more pythonic, but just sharing...

```
checks = ["RUNNING", "SUBMITTING", "QUEUED"]
if any(status in checks for status in statuses):
    ....
```

christinaholtNOAA commented on tests/WE2E/monitor_jobs.py:
```diff
@@ -0,0 +1,280 @@
+    """
+
+    #If we are no longer tracking this experiment, return unchanged
+    if expt["status"] in ['DEAD','ERROR','COMPLETE']:
```
Is `expt` guaranteed to have a `status` entry here?

mkavulich replied:
It is unless the input dictionary is misformatted. When the dict is created (either after the input file is read or in `run_WE2E_tests()`) the status is always set.

christinaholtNOAA commented on tests/WE2E/monitor_jobs.py:
```diff
@@ -0,0 +1,280 @@
+    with closing(sqlite3.connect(db)) as connection:
+        with closing(connection.cursor()) as cur:
+            db = cur.execute(ex).fetchall()
+    return db
```
Any chance this is a problem if a user calls this script for a real-time experiment whose first cycle is a LONG time ago and the cycles since then are all inactive?

mkavulich replied:
This script is not designed with real-time experiments in mind, just the WE2E tests. If realtime functionality is desired and the current logic is insufficient then those considerations can be taken into account in future updates.

christinaholtNOAA commented on ush/setup.py:
```diff
@@ -152,9 +158,6 @@ def load_config_for_setup(ushdir, default_config, user_config):
 
-    for k, v in cfg_d["task_run_fcst"].items():
-        print(f"*** {k}: {v}")
-
```
Eek! Sorry.

christinaholtNOAA reviewed (commented):
@mkavulich I reviewed all the files _except_ the main show (run_WE2E_tests.py)...I should be able to come back to that a little later. Sorry to break it up into sections.

christinaholtNOAA commented on tests/WE2E/run_WE2E_tests.py:
```diff
@@ -0,0 +1,447 @@
+    setup_logging(debug=args.debug)
+
+    # Set some important directories
+    ushdir=homedir + '/ush'
```
`os.path.join` will help with platform portability issues.

christinaholtNOAA commented on tests/WE2E/run_WE2E_tests.py:
```diff
@@ -0,0 +1,447 @@
+            continue
+        match=False
+        # Search for exact config file name to avoid accidental partial matches
+        test_config='config.' + test.rstrip() + '.yaml'
```
Consider an f-string here. They're generally more readable.

Is there any chance that the `test` might have leading whitespace? Would `strip` be better?

christinaholtNOAA commented on tests/WE2E/run_WE2E_tests.py:
```diff
@@ -0,0 +1,447 @@
+                tests_to_run.remove(testfile)
+    if len(tests_to_run) != len(set(tests_to_run)):
+        logging.warning("\nWARNING: Duplicate test names were found in list. Removing duplicates and continuing.\n")
+        tests_to_run = list(set(tests_to_run))
```
These two checks are _real_ nice!

christinaholtNOAA commented on tests/WE2E/run_WE2E_tests.py:
```diff
@@ -0,0 +1,447 @@
+            else:
+                # If we have gotten this far then the only option left for user_spec_tests is a file containing test names
+                if os.path.isfile(user_spec_tests[0]):
+                    tests_to_check = list(open(user_spec_tests[0]))
```
Where does this get closed?

mkavulich replied:
Allegedly (via the wisdom of stackoverflow) files are automatically closed when they go out of scope, and here the scope is only within the call to `list()`. However, I can't find any explicit reference to this in the official python documentation, so I have changed to a two-liner `with` statement.

christinaholtNOAA commented on tests/WE2E/run_WE2E_tests.py:
```diff
@@ -0,0 +1,447 @@
+            _ = check_tests(user_spec_tests)
+            # If no error, user_spec_tests is the name of the test specified on the command line
+            tests_to_check = user_spec_tests
+        except:
```
From a more purist perspective, using a try/except block is really meant to catch and handle exceptional situations. In this case, it seems it's being used to control logic. While that's not completely un-allowed, a blind except here does not meet the PEP-8 standard for good reason.

What if a user misspells a test name, but had every intent of doing this option correctly? I believe we'd get no useful information from the try/except block, and it would go along attempting to process the input as an attempt at specifying a file name and fail.

It may be helpful to check to see if a provided a file name...does it contain ".yaml", for example? Is one of the key words? Then, is it a test name?

mkavulich replied:
Re-arranged to get rid of "impure" `try` block 😉

christinaholtNOAA commented on tests/WE2E/run_WE2E_tests.py:
```diff
@@ -0,0 +1,447 @@
+                tests_to_check = []
+                for f in alltests:
+                    filename = os.path.basename(f)
+                    filename = filename[7:-5]
```
Why `[7:-5]`? I assume it's because `config*.yaml`, but maybe a comment would be helpful?

christinaholtNOAA commented on tests/WE2E/run_WE2E_tests.py:
```diff
@@ -0,0 +1,447 @@
+                    filename = filename[7:-5]
+                    tests_to_check.append(filename)
+                logging.debug(f"Will check all tests:\n{tests_to_check}")
+            elif user_spec_tests[0] == 'fundamental' or user_spec_tests[0] == 'comprehensive':
```
Shorter option:
```
elif user_spec_tests[0] in ['fundamental', 'comprehensive']:
```

christinaholtNOAA commented on tests/WE2E/run_WE2E_tests.py:
```diff
@@ -0,0 +1,447 @@
+                    tests_to_check.append(filename)
+                logging.debug(f"Will check all tests:\n{tests_to_check}")
+            elif user_spec_tests[0] == 'fundamental' or user_spec_tests[0] == 'comprehensive':
+                # I am writing this section of code under protest; we should use args.run_envir to check for run_envir-specific files!
```
I understand your concern!

While it's hideous, you may consider reducing duplication by setting a variable early on to set the "base file path" and then use that to build up your many options of that file name. If the files were all moved to a different subdir, for example, that'd be a LOT of edits to make instead of one.

christinaholtNOAA commented on tests/WE2E/run_WE2E_tests.py:
```diff
@@ -0,0 +1,447 @@
+    logging.info(f'Will run {len(tests_to_run)} tests:\n{pretty_list}')
+
+
+    config_default_file = ushdir + '/config_defaults.yaml'
```
os.path.join() here would be better. And `machine_file` below.

christinaholtNOAA commented on tests/WE2E/run_WE2E_tests.py:
```diff
@@ -0,0 +1,447 @@
+    # If RUN_TASK_GET_EXTRN_ICS is false, do nothing and return
+    if 'workflow_switches' in cfg:
+        if 'RUN_TASK_GET_EXTRN_ICS' in cfg['workflow_switches']:
+            if cfg['workflow_switches']['RUN_TASK_GET_EXTRN_ICS'] is False:
```
We don't often have a bit of logic that is necessarily this wordy. Perhaps these lines can be simplified?

```
if not cfg['workflow_switches'].get('RUN_TASK_GET_EXTRN_ICS', True):
    return cfg_ics
```

This logic says, if it's set and False, return. If it's not there, keep going. If it's there and set to True keep going.

mkavulich replied:
The example you give will fail if the key `workflow_switches` is not set, and it will return if RUN_TASK_GET_EXTRN_ICS is not set, neither of which is acceptable. Unfortunately because we can't guarantee that this key exists, *and* we need to check that *if* the key cfg['workflow_switches']['RUN_TASK_GET_EXTRN_LBCS'] is set it is *false* (if it is not set, it defaults to `True`), this logic necessarily needs to be a little convoluted.

christinaholtNOAA replied:
Oh. Sure. What about:

```
if cfg.get('workflow_switches', {}).get('RUN_TASK_GET_EXTRN_ICS', True) is False:
    return cfg_ics
```

The logic should return if `workflow_switches` is set and `RUN_TASK_GET_EXTRN_ICS` is set to something that evaluates to False.

If `workflow_switches` or `RUN_TASK_GET_EXTRN_ICS` isn't set, or `RUN_TASK_GET_EXTRN_ICS` evaluates to something not False, it does not return.

christinaholtNOAA commented on tests/WE2E/run_WE2E_tests.py:
```diff
@@ -0,0 +1,447 @@
+        return cfg_ics
+    if not cfg_ics['USE_USER_STAGED_EXTRN_FILES']:
+        logging.debug(f'USE_USER_STAGED_EXTRN_FILES is false for task_get_extrn_ics section of config')
+        return cfg_ics
```
Similar thing here...if it doesn't really matter to the user if it's not set or False:

```
if not cfg_ics.get('USE_USER_STAGED_EXTRN_FILES'):
    return cfg_ics
```

This ability to check if a key exists may also potentially be helpful in cleaning up the logic below surrounding `EXTRN_MDL_SYSBASEDIR_ICS`.

mkavulich replied:
Simplified logic using `.get()` method here and in lbcs function

christinaholtNOAA commented on tests/WE2E/run_WE2E_tests.py:
```diff
@@ -0,0 +1,447 @@
+    return cfg_ics
+
+def check_task_get_extrn_lbcs(cfg: dict, mach: dict, dflt: dict) -> dict:
+    """
```
This is a big copy with some changes from ICS to LBCS!

mkavulich replied:
Consolidated more duplicate code into `update_expt_status()`

christinaholtNOAA reviewed (commented):
This looks good to me! Only a handful of comments about minor changes you may consider.

mkavulich commented:
@gsketefian to answer your question here, yes, this script does solve the issue of the current script omitting invalid options silently. This new script takes a different approach to constructing the config.yaml file from the provided test one, in that we only perform the necessary manipulations to test the features needed and do not attempt to rebuild config.yaml from scratch. So if there is an invalid option provided, that will be passed through to the `generate_FV3LAM_workflow()` function, where the error will be caught and fail.

As an aside, I have documented this issue in #568, so it can be properly addressed

gsketefian commented on tests/WE2E/run_WE2E_tests.py:
```diff
@@ -0,0 +1,447 @@
+        if not match:
+            raise Exception(f"Could not find test {test}")
+    # Because some test files are symlinks to other tests, check that we don't
+    # include the same test twice
```
@mkavulich I thought you were getting rid of the symlinks feature.  Seems like you're still checking for that here.

mkavulich replied:
Sorry if I was confusing in our previous conversation; I had no plans to change the existence of symlinks; I only changed it so that this does not result in a failure, it simply omits duplicate tests. Whether or not that's the best approach, I'm all ears on opinions there.

gsketefian commented on tests/WE2E/run_WE2E_tests.py:
```diff
@@ -0,0 +1,447 @@
+    optional = parser._action_groups.pop() # Create a group for optional arguments so they can be listed after required args
+    required = parser.add_argument_group('required arguments')
+
+    required.add_argument('-m', '--machine', type=str, help='Machine name; see ush/machine/ for valid values', required=True)
```
@mkavulich The "machine" argument needs to be made lower-case before being used since this failed:
```
./run_WE2E_tests.py --machine Hera --account rtrr --tests MET_verification --expt_basedir set01
```
with
```
FileNotFoundError: [Errno 2] No such file or directory: '/scratch2/BMC/det/Gerard.Ketefian/UFS_CAM/TEST_mkavulich_new_run_WE2E_tests/ufs-srweather-app/ush/machine/Hera.yaml'
```
but changing "Hera" to "hera" worked.

mkavulich replied:
Fixed, this was needed a few other places too.

gsketefian commented:
@mkavulich I tried testing what happens with duplicate file names:  I copied `wflow_features/config.MET_verification.yaml` to a new directory under `test_configs` named `verification`, i.e. I created the file `verification/config.MET_verification.yaml`.  The files are identical.  

When I launched `run_WE2E_tests.py`, it tried running both instances, so it would be good to have it check for duplicate tests under the `test_configs` directory.  In this case, one of the `MET_verification` experiment directories got renamed to `MET_verification_old_...` (probably when the 2nd one launched), and the script seemed to monitor one experiment thereafter (I guess the first test failed because of the directory rename, but I'm not sure).

mkavulich committed:
Add check that there are no duplicate test file names

mkavulich commented:
@gsketefian The case of duplicate file names is not one I had accounted for; I had only accounted for the user specifying duplicate tests. I just pushed a new commit with a check that there are no config.TESTNAME.yaml files with the same TESTNAME.

mkavulich committed:
Consolidate rocotorun_cmd into a defined variable for clarity/compactness

mkavulich committed:
Fix bug with debug flag causing pernicious I/O errors

mkavulich committed:
Format improvement suggested by Christina

mkavulich committed:
Merge commit '25418e86315c1df1e20ca2fd7e096124c28efec4' into feature/issue_462_new_WE2E_script_rebased

mkavulich committed:
Address more review comments:

 - Use os.path.join for platform agnosticism
 - Rearrange logic to avoid use of non-failure use of "try" block
 - Fix another bug if blank lines are included in test file
 - Put un-closed calls to open() in "with" block

mkavulich committed:
Convert --machine argument to lowercase, use os.path.join for paths

mkavulich committed:
Forgot to commit this fix for bad directory name

mkavulich committed:
In monitor_jobs.py:

 - Expand update_expt_status() to reduce duplicate code
 - Update docstrings

mkavulich committed:
Per review comments, use .get() method in dictionaries to simplify logic

mkavulich committed:
Address more review comments

mkavulich committed:
Address final (?) reviewer comments

mkavulich commented:
@gsketefian @christinaholtNOAA Thank you for your reviews, I believe I have addressed all your concerns. Let me know if I missed something or you have more questions/comments.

gsketefian commented:
@mkavulich I was testing redundancy with symlinks (I had a symlink to a test, and I listed both in my list of tests), and the script caught the redundancy with an appropriate warning.  Just the output to screen is not well-formatted (seems to be a lot of extra spaces where there should be just one) and thus a bit hard to read.  Do you mind fixing if possible?  Here's what the output looked like:
```
$ ./run_WE2E_tests.py --machine hera --account rtrr --tests my_tests.txt --expt_basedir /scratch2/BMC/det/Gerard.Ketefian/UFS_CAM/TEST_mkavulich_new_run_WE2E_tests/expt_dirs/set01 -q
Checking that all tests are valid
WARNING: test file /scratch2/BMC/det/Gerard.Ketefian/UFS_CAM/TEST_mkavulich_new_run_WE2E_tests/ufs-srweather-app/tests/WE2E/test_configs/verification/config.MET_verification_link.yaml is a symbolic link to a
                                test file (/scratch2/BMC/det/Gerard.Ketefian/UFS_CAM/TEST_mkavulich_new_run_WE2E_tests/ufs-srweather-app/tests/WE2E/test_configs/wflow_features/config.MET_verification.yaml) that is also included in the
                                test list. Only the latter test will be run.
Will run 1 tests:
/scratch2/BMC/det/Gerard.Ketefian/UFS_CAM/TEST_mkavulich_new_run_WE2E_tests/ufs-srweather-app/tests/WE2E/test_configs/wflow_features/config.MET_verification.yaml
Workflow for test MET_verification successfully generated in
/scratch2/BMC/det/Gerard.Ketefian/UFS_CAM/TEST_mkavulich_new_run_WE2E_tests/expt_dirs/set01/MET_verification

calling function that monitors jobs, prints summary
Writing information for all experiments to monitor_jobs_20230202091032.yaml
Checking tests available for monitoring...
Starting experiment MET_verification running
Setup complete; monitoring 1 experiments
```

gsketefian commented:
@mkavulich Two questions came to mind as I was doing further tests (these are just for discussion; I'm approving the PR):
1) The current (shell) version searches only a specific set of subdirectories under `test_config` for tests.  I had done it that way because often during testing I may make a backup copy of a subdirectory before starting to modify the files in it.  The new script searches all subdirectories.  I'm ok with that (I can always copy backup directories to somewhere outside of `test_configs`), just wondering how you decided on that.  Just code simplicity?
2) Once the cron features are enabled in a future PR, will the monitoring script be able to monitor tasks running with cron as well?  Currently, there's a script (`get_expts_status.sh`) that does this.  It would be nice to have a proper replacement for it (or just leave it as is if not).

christinaholtNOAA reviewed (approved):
As I mentioned last week, I'm totally fine with the PR as-is, but left just a couple of follow up comments just to circle back to the review.

mkavulich commented:
Thanks @christinaholtNOAA; for expediency and to avoid unnecessary testing, I will roll these suggestions into my working branch for the next update to these scripts.
